File: main/scala/photos/brooklyn/practice/codingexercises/scala/leetcode/solutions/IncreasingOrderSearchTree.scala
package photos.brooklyn.practice.codingexercises.scala.leetcode.solutions

object IncreasingOrderSearchTree {

  def increasingBST(root: TreeNode, tail: TreeNode=null): TreeNode =
    if (root == null) return tail
    else {
      val res = increasingBST(root.left, root)
      res.left = null
      res.right = increasingBST(root.right, tail)
      res
    }


}

class TreeNode(var _value: Int) {
  var value: Int = _value
  var left: TreeNode = null
  var right: TreeNode = null
}

File: main/java/photos/brooklyn/practice/codingexercises/java/leetcode/solutions/IncreasingOrderSearchTree.java
package photos.brooklyn.practice.codingexercises.java.leetcode.solutions;

/**
 * https://leetcode.com/problems/increasing-order-search-tree/discuss/165885/C++JavaPython-Self-Explained-5-line-O(N)
 */
public class IncreasingOrderSearchTree {
    public TreeNode increasingBST(TreeNode root) {
        return helper(root, null);
    }

    public TreeNode helper(TreeNode root, TreeNode tail) {
        if (root == null) return tail;
        TreeNode res = helper(root.left, root);
        root.left = null;
        root.right = helper(root.right, tail);
        return res;
    }
}

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}
